### PR TITLE
feat: Support Chunked Linear Fusion for GRPO

### PR DESCRIPTION
**Pull request by pengdurice**
# What does this PR do ?

**Support Linear CE Loss Fusion for GRPO.**

On top of #2036 (SFT) and #2139 (DPO), this PR extends the chunked linear
cross-entropy fusion loss to GRPO (and the other policy-gradient variants that
share `ClippedPGLossFn`: PPO, REINFORCE/RLOO, GSPO, Dr.GRPO, etc.).

# Optimizations

## Chunked Linear Cross-Entropy Fusion Loss

During standard GRPO training the model materializes a full logit tensor of
shape `[batch_size, seq_length, vocab_size]` for the policy forward-backward
pass and for the previous-/reference-policy logprob computations. This can
cause out-of-memory (OOM) errors for long sequences or large vocabularies. The
**chunked linear cross-entropy fusion loss** avoids this by computing the
per-token log probabilities directly from the hidden states: it chunks the
sequence dimension, projects each chunk to logits on the fly, gathers the
realized-token log probabilities, and discards the logits before moving to the
next chunk.

**Benefits:**

- Extends the maximum trainable sequence length by eliminating the large logit
  tensor from GPU memory.
- Applies to the training forward-backward pass **and** the previous/reference
  policy logprob computations (the worker `get_logprobs` path already honors the
  flag).
- Produces numerically equivalent loss/reward dynamics to the standard path.

# Key changes

1. `ClippedPGLossFn.__init__` now accepts `use_linear_ce_fusion`. The generic
   `prepare_loss_input` already treats the model output as precomputed
   next-token logprobs for any `LossInputType.LOGPROB` loss whose
   `use_linear_ce_fusion` attribute is set, so no per-loss branching was needed.
2. `nemo_rl/algorithms/grpo.py` reads
   `policy.megatron_cfg.use_linear_ce_fusion_loss`, passes it into
   `ClippedPGLossFn`, and adds guardrails (see Constraints below).
3. `nemo_rl/models/megatron/setup.py`: when fusion is enabled, force
   `overlap_param_gather` off (with a warning) and skip the
   forward-pre-hook-disable path. The fused forward runs the decoder but reads
   `output_layer.weight` directly instead of calling `output_layer.forward()`,
   which breaks Megatron's distributed-optimizer param-gather prefetch chain
   (`assert self.param_gather_handle is None` in `param_and_grad_buffer.py`).
   This is most easily triggered on MoE models with expert parallelism.
4. `nemo_rl/distributed/model_utils.py`: the patched GPT forward now accepts and
   forwards the newer Megatron-Core `is_spec_decode` kwarg (only when provided),
   keeping the monkey-patch in sync with the current `GPTModel.forward`
   signature while staying compatible with older ones.

# Constraints

Enforced in `nemo_rl/algorithms/grpo.py` / `setup.py`:

- Megatron backend only (`policy.megatron_cfg.enabled: true`).
- Context parallelism must be 1.
- Sequence packing must be disabled (the fused forward rolls labels over the
  whole sequence and would mix tokens across packed boundaries).
- Top-k/top-p training-time filtering must be off (`top_k: null`, `top_p: 1.0`),
  since the fused path gathers logprobs from unfiltered logits.

# Issues

NA

# Tests

1. **Unit test** — `test_megatron_grpo_linear_ce_fusion_agreement` in
   `tests/unit/models/policy/test_megatron_worker.py` compares the standard
   `ClippedPGLossFn` loss against the fusion-enabled loss on a tiny Qwen2 model
   (rtol/atol = 1e-2), mirroring the SFT/DPO agreement tests.
2. **Nightly functional test** — added
   `tests/test_suites/llm/grpo-qwen2.5-math7b-1n8g-megatron_chunked_linear_ce_loss.sh`
   + recipe under `examples/configs/recipes/llm/` and registered it in
   `tests/test_suites/nightly.txt`.
3. **Large-scale A/B convergence check** — GRPO on **Qwen3-30B-A3B** (MoE,
   Megatron TP=2 × EP=8 on 2×8 H200, vLLM TP=4 colocated, seq 4096), 50 steps,
   fusion ON vs OFF, two seeds each. Configs are identical except the fusion
   flag.

   | run | fusion | mean reward (50 steps) | gen-KL error |
   | --- | --- | --- | --- |
   | A1 | on  | 0.523 | ~0.002 |
   | A2 | on  | 0.524 | ~0.002 |
   | B1 | off | 0.518 | ~0.002 |
   | B2 | off | 0.518 | ~0.002 |

   - Step 1 loss/reward are identical across all runs (generation is
     deterministic from the seed and the optimizer has not yet diverged),
     confirming the fused logprobs/gradients match the standard path.
   - The train-vs-inference logprob error (`Generation KL Error`) is ~0.002 and
     step-for-step identical between fusion and baseline.
   - Aggregate reward differs by <1% (within run-to-run RL noise — the two
     fusion runs differ by more than fusion-vs-baseline does), and the four
     reward/loss trajectories interleave with no systematic gap. No
     instability/NaN; all runs completed 50/50 steps.

# Usage

Enable `megatron_cfg` and add the two flags below to your GRPO config:

```yaml
policy:
  megatron_cfg:
    enabled: true
    use_linear_ce_fusion_loss: true
    linear_ce_fusion_chunk_size: 256  # tokens per chunk; smaller = less memory, larger = more throughput
  # required with fusion:
  sequence_packing:
    enabled: false
  generation:
    top_k: null
    top_p: 1.0
```

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed Contributor guidelines
- [x] Did you write any new necessary tests?
- [x] Did you run the unit tests and functional tests locally?
- [x] Did you add or update any necessary documentation? (docs/guides/grpo.md)

# Additional Information

- Docs: added a "Chunked Linear Cross-Entropy Fusion Loss" subsection to
  `docs/guides/grpo.md`.
- The bulk of the fusion machinery is shared infrastructure introduced in #2036
  and is algorithm-agnostic; this PR is mostly wiring + guardrails + the
  distributed-optimizer `overlap_param_gather` fix surfaced by MoE + EP.

### CODE REVIEW

File: nemo_rl/models/generation/vllm/vllm_generation.py [hunk 1]
```diff
@@ -169,6 +169,13 @@ def __init__(
 
         # It's necessary to set env_vars here to ensure that vllm non-leader workers also have these env_vars
         env_vars = {}
+        # vLLM 0.20 auto-selects the FlashInfer CUTLASS/TRTLLM fused-MoE backend for
+        # MoE models (e.g. Qwen3-30B-A3B). Those kernels produce incorrect outputs
+        # (0% accuracy / gibberish) for several Qwen3 MoE configs, see vLLM issues
+        # #34892, #37591, #37758. Force the reference Triton MoE backend instead.
+        # (Set before the per-recipe env_vars loop so a yaml can still override.)
+        env_vars["VLLM_USE_FLASHINFER_MOE_FP16"] = "0"
```
Comment (yuki-97) [inline nemo_rl/models/generation/vllm/vllm_generation.py]: can we move the two env_vars to `examples/configs/recipes/llm/grpo-qwen3-30ba3b-2n8g-megatron_chunked_linear_ce_loss.yaml`? so that other configs won't be affect.

File: nemo_rl/models/megatron/setup.py [hunk 2]
```diff
@@ -894,6 +894,26 @@ def _create_megatron_config(
...
+    if (
+        config["megatron_cfg"].get("use_linear_ce_fusion_loss", False)
+        and overlap_param_gather
+    ):
+        warnings.warn(
+            "Disabling overlap_param_gather because linear CE fusion loss is enabled: "
+            "the fused forward bypasses output_layer.forward() and is incompatible "
+            "with the distributed-optimizer param-gather prefetch chain."
+        )
+        overlap_param_gather = False
```
Comment (yuki-97) [inline nemo_rl/models/megatron/setup.py]: can we just assert False when using use_linear_ce_fusion_loss + overlap_param_gather?

File: nemo_rl/models/megatron/setup.py [hunk 3]
```diff
@@ -1486,6 +1506,9 @@ def finalize_megatron_setup(
         and config["megatron_cfg"]["distributed_data_parallel_config"][
             "overlap_param_gather"
         ]
+        # Linear CE fusion forces overlap_param_gather off (see
+        # create_megatron_config), so there is no forward pre-hook to disable.
+        and not config["megatron_cfg"].get("use_linear_ce_fusion_loss", False)
```
Comment (yuki-97) [inline nemo_rl/models/megatron/setup.py]: here seems can be removed if this done. https://github.com/NVIDIA-NeMo/RL/pull/2833/changes#r3421595432

File: nemo_rl/algorithms/grpo.py [hunk 4]
```diff
@@ -408,7 +412,41 @@ def init_train_dataloader(dataset, suffix: str = ""):
...
     #        Loss Function
     # ==========================
-    loss_fn = ClippedPGLossFn(loss_config)
+    # Linear CE fusion loss computes next-token logprobs directly from hidden states
+    # (chunked over the sequence) and never materializes the full
+    # [batch, seq_len, vocab_size] logit tensor, which significantly reduces peak
+    # memory. It is only available on the Megatron backend.
+    use_linear_ce_fusion = policy_config["megatron_cfg"]["enabled"] and policy_config[
+        "megatron_cfg"
+    ].get("use_linear_ce_fusion_loss", False)
```
Comment (yuki-97) [inline nemo_rl/algorithms/grpo.py]: nit
```suggestion
    use_linear_ce_fusion = (
        policy_config["megatron_cfg"]["enabled"]
        and policy_config["megatron_cfg"]["use_linear_ce_fusion_loss"]
    )
```

**Review by yuki-97 — COMMENTED** [review]
@pengdurice thanks for adding this to GRPO as well, overall LGTM and left some minor comments.

File: examples/configs/grpo_math_1B.yaml [hunk 5]
```diff
@@ -144,6 +144,13 @@ policy:
   
   megatron_cfg:
     enabled: false
+    # Compute next-token logprobs directly from hidden states with a chunked
+    # linear cross-entropy fusion, avoiding materialization of the full
+    # [batch, seq_len, vocab_size] logit tensor. Reduces peak memory and extends
+    # the maximum trainable sequence length. Not compatible with context
+    # parallelism, sequence packing, or top-k/top-p training-time filtering.
+    use_linear_ce_fusion_loss: false
```
Comment (yuki-97) [inline examples/configs/grpo_math_1B.yaml]: feels `ce_fusion_loss` is a bit confused here since GRPO is not using CE.
is there a better name for it? and could also apply the new name to SFT and DPO to keep the same.

File: nemo_rl/models/generation/vllm/vllm_generation.py (hunk 1 above)
Comment (pengdurice) [reply nemo_rl/models/generation/vllm/vllm_generation.py]: yeah, good catch, changed!

File: nemo_rl/models/megatron/setup.py (hunk 3 above)
Comment (pengdurice) [reply nemo_rl/models/megatron/setup.py]: sg, removed!

File: nemo_rl/algorithms/grpo.py (hunk 4 above)
Comment (pengdurice) [reply nemo_rl/algorithms/grpo.py]: thank for this, since the name needs to be changed, so I didn't sign off and apply this suggestion directly, but thank you for pointing this out;)

File: examples/configs/grpo_math_1B.yaml (hunk 5 above)
Comment (pengdurice) [reply examples/configs/grpo_math_1B.yaml]: yes, I changed to use_fused_linear_logprobs which is more general, does that make sense? Also for DPO, it is using the logprobs similar to GPRO. so I also changed there. LMK if that makes sense;-) thank you for pointing this out!

Comment (yuki-97) [reply examples/configs/grpo_math_1B.yaml]: thanks for updating this! and the name of yaml and script may also need updates. others LGTM.
```
tests/test_suites/llm/sft-qwen2.5-math7b-1n8g-megatron_chunked_linear_ce_loss.sh
tests/test_suites/llm/dpo-qwen2.5-math7b-1n8g-megatron_chunked_linear_ce_loss.sh
tests/test_suites/llm/grpo-qwen3-30ba3b-2n8g-megatron_chunked_linear_ce_loss.sh
```

File: nemo_rl/models/megatron/setup.py [hunk 6]
```diff
@@ -953,6 +953,21 @@ def _create_megatron_config(
...
 
+    # Fused linear logprobs run the decoder but read output_layer.weight directly
+    # instead of calling output_layer.forward(). Megatron's distributed-optimizer
+    # overlap_param_gather prefetch chain assumes every param-gather bucket
+    # (including the output layer) is consumed by a module forward; skipping it
+    # leaves a stale param_gather_handle and trips
+    #   assert self.param_gather_handle is None  (param_and_grad_buffer.py)
+    # on the next iteration, so the two are mutually exclusive.
+    if config["megatron_cfg"].get("use_fused_linear_logprobs", False):
+        assert not overlap_param_gather, (
```
Comment (yuki-97) [inline nemo_rl/models/megatron/setup.py:962]: hi @pengdurice , thanks for the update!

I found both `sft-qwen2.5-math7b-1n8g-megatron_fused_linear_logprobs` and `dpo-qwen2.5-math7b-1n8g-megatron_fused_linear_logprobs` will meet the assert here, could you help to check?

File: examples/configs/recipes/llm/grpo-qwen3-30ba3b-2n8g-megatron_fused_linear_logprobs.yaml [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+      overlap_param_gather: false
+  generation:
+    vllm_cfg:
+      env_vars:
+        VLLM_USE_FLASHINFER_MOE_FP16: '0'
+        VLLM_USE_FLASHINFER_MOE_FP8: '0'
+loss_fn:
+  reference_policy_kl_penalty: 0.0
+logger:
+  log_dir: logs/grpo-qwen3-30ba3b-2n8g-megatron_chunked_linear_ce_loss
```
Comment (yuki-97) [inline examples/configs/recipes/llm/grpo-qwen3-30ba3b-2n8g-megatron_fused_linear_logprobs.yaml]: also the logger fields inside these recipes still use `chunked-linear-ce-loss`:
- `grpo-...-fused_linear_logprobs.yaml:27,31,33,35`
- `dpo-...-fused_linear_logprobs.yaml:38,40,42`
- `sft-...-fused_linear_logprobs.yaml:48,50,52`

**Review by yuki-97 — COMMENTED** [review]
stale logger-field names after the rename

File: examples/configs/recipes/llm/grpo-qwen3-30ba3b-2n8g-megatron_fused_linear_logprobs.yaml (hunk 7 above)
Comment (pengdurice) [reply examples/configs/recipes/llm/grpo-qwen3-30ba3b-2n8g-megatron_fused_linear_logprobs.yaml]: thank you for pointing this out, fixed!

File: nemo_rl/models/megatron/setup.py (hunk 6 above)
Comment (pengdurice) [reply nemo_rl/models/megatron/setup.py:962]: Good catch — fixed by setting overlap_param_gather: false in both the sft and dpo fused_linear_logprobs recipes (mirroring the grpo recipe), so they satisfy the assert. Thanks!

**Review by yuki-97 — DISMISSED** [review]
LGTM, thanks @pengdurice

**Comment by pengdurice** [comment]
> LGTM, thanks @pengdurice

Thank you! Just fixed two issues that caused the error in the last CICD run.

**Review by yuki-97 — DISMISSED** [review]

**Comment by pengdurice** [comment]
> /ok to test [0e92d1c](https://github.com/NVIDIA-NeMo/RL/commit/0e92d1c5c5d395f185014ad3c69d739124645b92)

@yuki-97 , thank you for approving this! it seems that the GPU hour limitation is reached again. Would you mind lift the limit again? thank you!

**Comment by pengdurice** [comment]
> > /ok to test [0e92d1c](https://github.com/NVIDIA-NeMo/RL/commit/0e92d1c5c5d395f185014ad3c69d739124645b92)
> 
> @yuki-97 , thank you for approving this! it seems that the GPU hour limitation is reached again. Would you mind lift the limit again? thank you!

The limit for this CICD is 2300, however, most recent main is at 3270. Would you mind merge main / rebase main to it or should I do it? thank you!

**Comment by pengdurice** [comment]
> > > /ok to test [0e92d1c](https://github.com/NVIDIA-NeMo/RL/commit/0e92d1c5c5d395f185014ad3c69d739124645b92)
> > 
> > 
> > @yuki-97 , thank you for approving this! it seems that the GPU hour limitation is reached again. Would you mind lift the limit again? thank you!
> 
> The limit for this CICD is 2300, however, most recent main is at 3270. Would you mind merge main / rebase main to it or should I do it? thank you!

I reduced the training steps and required time for all 3 fused logprob jobs (and test locally again on h200 node) and could you help take a look again? thank you!

**Comment by yuki-97** [comment]
hi @pengdurice , thanks for fixing these! LGTM and I've started CI agin.

feel free to merge / rebase main and update the nightly limit yourself. we usually update the nightly time limit to current + 40 in order to avoid nightly scripts that have huge time increase.

**Review by yuki-97 — APPROVED** [review]

**Comment by pengdurice** [comment]
> hi @pengdurice , thanks for fixing these! LGTM and I've started CI agin.
> 
> feel free to merge / rebase main and update the nightly limit yourself. we usually update the nightly time limit to current + 40 in order to avoid nightly scripts that have huge time increase.

Thank you! Will do next time;-)
